### PR TITLE
[Master] Fix 13602 preserve comment following element in node list

### PR DESCRIPTION
**Pull request by yuit**
Fix issue #13602. 

- [x] https://github.com/Microsoft/TypeScript/issues/13602#issue-202246787

- [x] https://github.com/Microsoft/TypeScript/issues/13602#issuecomment-274219821

Wont fix at the moment
~~- [ ] https://github.com/Microsoft/TypeScript/issues/13602#issuecomment-274219821~~

~~- [ ] https://github.com/Microsoft/TypeScript/issues/13602#issuecomment-277914269~~

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -2274,6 +2286,20 @@ namespace ts {
...
+
+                // Emit any trailing comment of the last element in the list
+                // i.e
+                //       var array = [...
+                //          2
+                //          /* end of element 2 */
+                //       ];
+                if (previousSibling && delimiter && previousSibling.end !== parentNode.end) {
+                    emitLeadingComments(previousSibling.end, /*isEmittedNode*/ previousSibling.kind !== SyntaxKind.NotEmittedStatement);
+                    if (hasTrailingComma) {
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: Why are you emitting this twice? Also, statements are never emitted with delimiters, so your `isEmittedNode` argument will always be `true`.

File: tests/baselines/reference/commentLeadingCloseBrace.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        commentedParameters(5, 6);
+    }
+}
+
+//// [commentLeadingCloseBrace.js]
+function ifelse() {
+    if (commentedParameters(1, 2)) {
+        /*comment1*/
+        commentedParameters(3, 4);
+    /*comment2*/
```
Comment (rbuckton) [inline tests/baselines/reference/commentLeadingCloseBrace.js]: This should be indented similar to `/*comment1*/` above.

Comment (rbuckton) [inline tests/baselines/reference/commentLeadingCloseBrace.js]: We might want to investigate whether it makes sense to emit leading/trailing comments of a position in the `emitNodeList` function in `emitter.ts`.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -2228,6 +2231,15 @@ namespace ts {
 
                     // Write the delimiter if this is not the first node.
                     if (previousSibling) {
+                        // i.e
+                        //      function commentedParameters(
+                        //          /* Parameter a */
+                        //          a
+                        //          /* End of parameter a */ -> this comment doesn't consider to be trailing comment of parameter "a" due to newline
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: `doesn't consider` -> `isn't considered`

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -1346,6 +1347,8 @@ namespace ts {
             else {
                 writeToken(SyntaxKind.OpenBraceToken, node.pos, /*contextNode*/ node);
                 emitBlockStatements(node);
+                // We have to call emitLeadingComments explicitly here because otherwise leading comments of the close brace token will not be emitted
+                emitLeadingComments(node.statements.end, /*isEmittedNode*/true);
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: Perhaps we should emit leading/trailing comments of the node range in `emitNodeList`

File: tests/baselines/reference/argsInScope.js [hunk 5]
```diff
@@ -17,6 +17,7 @@ var C = (function () {
     }
     C.prototype.P = function (ii, j, k) {
         for (var i = 0; i < arguments.length; i++) {
+        // WScript.Echo("param: " + arguments[i]);
```
Comment (rbuckton) [inline tests/baselines/reference/argsInScope.js]: This should be indented.

File: tests/baselines/reference/amdImportAsPrimaryExpression.js [hunk 6]
```diff
@@ -26,5 +26,6 @@ define(["require", "exports"], function (require, exports) {
 define(["require", "exports", "./foo_0"], function (require, exports, foo) {
     "use strict";
     if (foo.E1.A === 0) {
+    // Should cause runtime import - interesting optimization possibility, as gets inlined to 0.
```
Comment (rbuckton) [inline tests/baselines/reference/amdImportAsPrimaryExpression.js]: This should be indented.

File: src/compiler/comments.ts [hunk 7]
```diff
@@ -32,6 +33,7 @@ namespace ts {
             emitNodeWithComments,
             emitBodyWithDetachedComments,
             emitTrailingCommentsOfPosition,
+            emitLeadingComments,
```
Comment (yuit) [inline src/compiler/comments.ts]: thanks @rbuckton ! I was wondering if I should create similar function call like `emitTrailingCommentsOfPosition`

File: src/compiler/emitter.ts (hunk 4 above)
Comment (yuit) [inline src/compiler/emitter.ts]: do you mean make a call like this `emitLeadingCommentsOfPosition(nodeList.pos)`?

File: tests/baselines/reference/commentLeadingCloseBrace.js (hunk 2 above)
Comment (yuit) [inline tests/baselines/reference/commentLeadingCloseBrace.js]: not sure I understand your comment. do you emit leading/trailing comments of each element in node-list

File: src/compiler/emitter.ts [hunk 8]
```diff
@@ -2228,6 +2233,15 @@ namespace ts {
 
                     // Write the delimiter if this is not the first node.
                     if (previousSibling) {
+                        // i.e
+                        //      function commentedParameters(
+                        //          /* Parameter a */
+                        //          a
+                        //          /* End of parameter a */ -> this comment isn't considered to be trailing comment of parameter "a" due to newline
+                        //          ,
+                        if (emitLeadingCommentsOfPosition && delimiter && previousSibling.end !== parentNode.end) {
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: There shouldn't be a need to test for the existence of `emitLeadingCommentsOfPosition`.

**Pull request merged**